### PR TITLE
Add initial SPDX 2.3 serializer and other improvements

### PR DESCRIPTION
**Pull request by puerco**
This PR adds the initial SPDX 2.3 serializer and renderer and piggybacks a bunch of other improvements:

* Initial SPDX 2.3 serializer
* Follow up changes to #3 : 
   * Renames pkg/format to pkg/formats to make it easy to keep reusing some previous code
   * Centralizes the CDX & SPDX hash algorithm resolution into the generated HashAlgorithm type
   * Refactors the CDX serializer to a single CycloneDX serializer which can be reused to support more versions

The CDX refactor in eb90cd8 minimizes the CycloneDX by abstracting  @houdini91 's code it into a common protobom → CDX serializer which can be embedded in each encoding+version type.

The SPDX serializer is not yet complete, but it already produces a usable SBOM using the official SPDX go libraries, here is a sample reading and re-writing the curl SBOM in the examples directory and piping it to [`bom`](https://github.com/kubernetes-sigs/bom):

image

Signed-off-by: Adolfo García Veytia (Puerco) <puerco@chainguard.dev>

### CODE REVIEW

**Comment by jspeed-meyers** [comment]
I added a number of reviewers, more or less randomly :) I say we try to get at least one review, if not two, and then  merge either later today or Monday morning ET at the latest (to keep the tempo up :) )

File: pkg/writer/serializer_cdx.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+		}
+	*/
+
+	// Generate all components
+	components := map[string]*cdx.Component{}
+	refless := []*cdx.Component{}
+	for _, n := range bom.Nodes {
+		comp := s.nodeToComponent(n)
+		if comp == nil {
+			// Erorr? Warrn?
```
Comment (jspeed-meyers) [inline pkg/writer/serializer_cdx.go:46]: silly nit: ignore, but `Error? Warn?`

File: pkg/writer/serializer_cdx.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+		// Add components without refs
+		for _, c := range refless {
+			*doc.Components = append(*doc.Components, *c)
+		}
+	}
+	return doc, nil
+}
+
+// nodeTo14Component converta a node in protobuf to a CycloneDX 1.4 component
```
Comment (jspeed-meyers) [inline pkg/writer/serializer_cdx.go:151]: `converta` -> `converts`

**Review by jspeed-meyers — APPROVED** [review]
Nice!

File: pkg/formats/spdx/spdx.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	if strings.HasPrefix(s, "Person:") {
+		actorType = "person"
+		s = strings.TrimPrefix(s, "Person:")
+	} else if strings.HasPrefix(s, "Organization:") {
+		actorType = "org"
+		s = strings.TrimPrefix(s, "Organization:")
+	}
+	s = strings.TrimSpace(s)
+	actorName = s
+	if strings.HasSuffix(s, ")") && strings.Contains(s, "(") {
```
[the remark this answers is not in the document]
Comment (puerco) [reply pkg/formats/spdx/spdx.go:31]: OK, added more documentation to the function:

```go
// ParseActorString parses an SPDX "actor string", it is a specially formatted
// string that contains the type of actor (Person/Organization), their name and
// optionally an email address. For example, the following string:
//
// "Person: Debian OpenLDAP Maintainers (pkg-openldap-devel@lists.alioth.debian.org)"
//
// would return:
// "Organization", "Debian OpenLDAP Maintainers", "pkg-openldap-devel@lists.alioth.debian.org"

```

**Comment by puerco** [comment]
OK, I'm pulling this in to keep going. I've fixed the nits in the comments :+1:

**Pull request merged**